tests/unit/spawn_test.cc

```diff
@@ -0,0 +1,113 @@
+    return process::spawn(false_cmd, {false_cmd}).then([] (auto process) {
+        return process.wait();
+    }).then([] (auto wstatus) {
+        auto* exit_status = std::get_if<process::wait_exited>(&wstatus);
+        BOOST_REQUIRE(exit_status != nullptr);
+        BOOST_CHECK_EQUAL(exit_status->code, EXIT_FAILURE);
+    });
+}
+
+SEASTAR_TEST_CASE(test_spawn_program_does_not_exist) {
+    const char* not_exist_cmd = "i/dont/exist";
+    return process::spawn(not_exist_cmd, {not_exist_cmd}).then_wrapped([] (future<process> fut) {
+        BOOST_REQUIRE(fut.failed());
+        BOOST_CHECK_EXCEPTION(std::rethrow_exception(fut.get_exception()),
+                              std::system_error,
+                              [](const auto& e) {
+                                  return e.code().value() == ENOENT;
+                              });
+    });
+}
+
+SEASTAR_TEST_CASE(test_spawn_echo) {
+    const char* echo_cmd = "/usr/bin/echo";
+    return process::spawn(echo_cmd, {echo_cmd, "-n", "hello", "world"}).then([] (auto process) {
+        return do_with(std::move(process), bool(true), [](auto& p, auto& matched) {
+            using consumption_result_type = typename input_stream<char>::consumption_result_type;
+            using stop_consuming_type = typename consumption_result_type::stop_consuming_type;
+            using tmp_buf = stop_consuming_type::tmp_buf;
+            struct consumer {
+                consumer(std::string_view expected, bool& matched)
+                    : _expected(expected), _matched(matched) {}
+                future<consumption_result_type> operator()(tmp_buf buf) {
+                    _matched = std::equal(buf.begin(), buf.end(), _expected.begin());
+                    if (!_matched) {
+                        return make_ready_future<consumption_result_type>(stop_consuming_type({}));
+                    }
+                    _expected.remove_prefix(buf.size());
+                    return make_ready_future<consumption_result_type>(continue_consuming{});
+                }
+                std::string_view _expected;
+                bool& _matched;
+            };
+            return p.stdout().consume(consumer("hello world", matched)).then([&matched] {
+                BOOST_CHECK(matched);
+            }).finally([&p] {
+                return p.wait().discard_result();
+            });
+        });
+    });
+}
+
+SEASTAR_TEST_CASE(test_spawn_input) {
+    const char* cat_cmd = "/usr/bin/cat";
+    static const sstring text = "hello world\n";
+    return process::spawn(cat_cmd, {cat_cmd}).then([] (auto process) {
+        return do_with(std::move(process), [](auto& p) {
+            return p.stdin().write(text).then([&p] {
+                return p.stdin().flush();
+            }).then([&p] {
+                return p.stdout().read_exactly(text.size());
+            }).then([] (temporary_buffer<char> echo) {
+                BOOST_CHECK(echo == echo.copy_of(text));
```

Comment:
could have used `BOOST_CHECK_EQ()` if we have `operator<<` for `temporary_buffer<char>`. will add it later.